modules/040-node-manager/hooks/yc_delete_preemptible_instances.go

```diff
@@ -0,0 +1,365 @@
+			Kind:                "NodeGroup",
+			FilterFunc:          applyNodeGroupFilter,
+		},
+	},
+}, deleteMachines)
+
+func deleteMachines(input *go_hook.HookInput) error {
+	var (
+		timeNow                        = time.Now().UTC()
+		machines                       []*Machine
+		preemptibleMachineClassesSet   = set.Set{}
+		nodeNameToNodeMap              = make(map[string]*Node)
+		nodeGroupNameToNodeGroupStatus = make(map[string]*NodeGroupStatus)
+	)
+
+	for _, mcRaw := range input.Snapshots["mcs"] {
+		if mcRaw == nil {
+			continue
+		}
+
+		ic, ok := mcRaw.(*YandexMachineClass)
+		if !ok {
+			return fmt.Errorf("failed to assert to *YandexMachineClass")
+		}
+
+		preemptibleMachineClassesSet.Add(ic.Name)
+	}
+
+	if preemptibleMachineClassesSet.Size() == 0 {
+		return nil
+	}
+
+	for _, nodeRaw := range input.Snapshots["nodes"] {
+		if nodeRaw == nil {
+			continue
+		}
+
+		node, ok := nodeRaw.(*Node)
+		if !ok {
+			return fmt.Errorf("failed to assert to *Node")
+		}
+
+		nodeNameToNodeMap[node.Name] = node
+	}
+
+	for _, ngStatusRaw := range input.Snapshots["nodegroupstatuses"] {
+		if ngStatusRaw == nil {
+			continue
+		}
+
+		ngStatus, ok := ngStatusRaw.(*NodeGroupStatus)
+		if !ok {
+			return fmt.Errorf("failed to assert to *NodeGroupStatus")
+		}
+
+		nodeGroupNameToNodeGroupStatus[ngStatus.Name] = ngStatus
+	}
+
+	for _, machineRaw := range input.Snapshots["machines"] {
+		machine, ok := machineRaw.(*Machine)
+		if !ok {
+			return fmt.Errorf("failed to assert to *Machine")
+		}
+
+		if machine.Terminating {
+			continue
+		}
+
+		if machine.MachineClassKind != "YandexMachineClass" {
+			continue
+		}
+
+		if !preemptibleMachineClassesSet.Has(machine.MachineClassName) {
+			continue
+		}
+
+		if node, ok := nodeNameToNodeMap[machine.Name]; ok {
+			machine.nodeCreationTimestamp = node.CreationTimestamp
+			machine.nodeGroup = node.NodeGroup
+		} else {
+			continue
+		}
+
+		machines = append(machines, machine)
+	}
+
+	if len(machines) == 0 {
+		return nil
+	}
+
+	for _, m := range getMachinesToDelete(timeNow, machines, nodeGroupNameToNodeGroupStatus) {
+		input.PatchCollector.Delete("machine.sapcloud.io/v1alpha1", "Machine", "d8-cloud-instance-manager", m)
+	}
+
+	return nil
+}
+
+func getMachinesToDelete(timeNow time.Time, machines []*Machine, ngToNgStatusMap map[string]*NodeGroupStatus) (machinesToDelete []string) {
+	sort.Slice(machines, func(i, j int) bool {
+		return machines[i].nodeCreationTimestamp.Before(&machines[j].nodeCreationTimestamp)
+	})
+
+	// we spread deletion over preemptibleDeletionPeriod partitioned by hookExecutionSchedule
+	// that is the calculated batch for Machines that are older than 20 hours, but younger than 24 hours
+	batch := len(machines) / int(preemptibleDeletionPeriod/hookExecutionSchedule)
+	if batch == 0 {
+		batch = 1
+	}
+
+	var criticallyOldMachinesToDelete []string
+	const durationThresholdForDeletion = preemptibleInstanceMaxLifetime - preemptibleDeletionPeriod
```

Comment:
This constant can be moved out of function because `preemptibleInstanceMaxLifetime` and `preemptibleDeletionPeriod` are both on top.

Reply:
Done.